core/include/traccc/geometry/detector_description.hpp

```diff
@@ -0,0 +1,240 @@
+    ///
+    /// @return A (const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    const auto& reference_y() const { return BASE::template get<5>(); }
+
+    /// Pitch for local position calculation in X direction (non-const)
+    ///
+    /// The position of a detector element (pixel or strip) is calculated
+    /// along the X axis with the formula:
+    /// @f$pos_x = reference_x + pitch_x * index_x@f$
+    ///
+    /// @return A (non-const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    auto& pitch_x() { return BASE::template get<6>(); }
+    /// Pitch for local position calculation in X direction (const)
+    ///
+    /// The position of a detector element (pixel or strip) is calculated
+    /// along the X axis with the formula:
+    /// @f$pos_x = reference_x + pitch_x * index_x@f$
+    ///
+    /// @return A (const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    const auto& pitch_x() const { return BASE::template get<6>(); }
+
+    /// Pitch for local position calculation in Y direction (non-const)
+    ///
+    /// The position of a detector element (pixel or strip) is calculated
+    /// along the Y axis with the formula:
+    /// @f$pos_y = reference_y + pitch_y * index_y@f$
+    ///
+    /// @return A (non-const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    auto& pitch_y() { return BASE::template get<7>(); }
+    /// Pitch for local position calculation in Y direction (const)
+    ///
+    /// The position of a detector element (pixel or strip) is calculated
+    /// along the Y axis with the formula:
+    /// @f$pos_y = reference_y + pitch_y * index_y@f$
+    ///
+    /// @return A (const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    const auto& pitch_y() const { return BASE::template get<7>(); }
+
+    /// The dimensionality (1D/2D) of the detector module (non-const)
+    ///
+    /// @return A (non-const) vector of @c char objects
+    ///
+    TRACCC_HOST_DEVICE
+    auto& dimensions() { return BASE::template get<8>(); }
+    /// The dimensionality (1D/2D) of the detector module (const)
+    ///
+    /// @return A (const) vector of @c char objects
+    ///
+    TRACCC_HOST_DEVICE
+    const auto& dimensions() const { return BASE::template get<8>(); }
+
+    /// The variance in the Y direction for 1D sensors (non-const)
+    ///
+    /// @return A (non-const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    auto& variance_y() { return BASE::template get<9>(); }
+    /// The variance in the Y direction for 1D sensors (const)
+    ///
+    /// @return A (const) vector of @c traccc::scalar objects
+    ///
+    TRACCC_HOST_DEVICE
+    const auto& variance_y() const { return BASE::template get<9>(); }
+
+    /// @}
+
+};  // class detector_description_interface
+
+/// SoA container describing the detector
+using detector_description = vecmem::edm::container<
```

Comment:
I have something just a little different in mind. :thinking:

As we discussed in the past, this type (currently called `traccc::detector_description`) will only be needed by the "first half" of the algorithms. Since track finding and fitting, as long as the measurements are set up correctly to support all different types of measurements in a single type, those algorithms should not care exactly how the measurements were created.

However I don't want to do things in the same generic way for the clusterization and measurement/spacepoint formation algorithms. Right now all our algorithms were written for pixel detectors. But in the near future I'll want to rename things such that it would be clear that all these are meant just for pixels. And then we would introduce a simpler clusterization algorithm for strips, and whatever other algorithms we would need to make wire chamber measurements.

In keeping with this code organization (where the client code would need to explicitly run pixel or strip clusterization on the appropriate containers), I imagine that we'll want to split the currently single `traccc::detector_description` class into a couple of different ones. Just because the detector description of pixels and strips could potentially still remain in a single class, we could either rename the current class to `traccc::sillicon_detector_description`, or possibly split it into a `traccc::pixel_detector_description` and `traccc::strip_detector_description` classes. And at one point we would introduce a `traccc::wire_detector_description` class.

Point being, the algorithms will only ever want to deal with one of these at a time. :thinking: At least I think so. So I wouldn't bundle them together. :thinking:

Reply:
Well I still think bundling would be better. Imagine a case that we call `read_detector()` for a detector with a combination of silicon and non-silocon (e.g. ATLAS Phase-I and Belle II), and other case with a detector only with either silicon (e.g. ATLAS Phase-II) or non-silicon (e.g. COMET, Mu2e). If they are not bundled, we will have to define function for every case... 

The clusterization won't take non-silicon container but this cannot be the reason to reject the bundling... you can just put `detector_description.silicon_container` into the clusterization algorithm. 

> Right now all our algorithms were written for pixel detectors.

Not quite. Tracking wire measurement is more production ready than that. We just don't have a detector description to use as an example. For example there is a running example for drift chamber with fitting: https://github.com/acts-project/traccc/blob/main/tests/cpu/test_kalman_fitter_wire_chamber.cpp

Reply:
:confused: The different ATLAS phases will be on the Athena code to handle. We will **not** be using the same (Athena) algorithm for running clusterization / measurement creation for the different periods. This is what a framework is for. That it detects what sort of input data a job has to process, and then schedules the appropriate algorithm to make it happen.

With the **toolset** that Acts / traccc is meant to be, I do want to have separate classes for doing these different things. I do not want to have a single algorithm called `traccc::cuda::clusterization_algorithm` in the long term. I'll want to have one called `traccc::cuda::pixel_clusterization_algorithm`, another one called `traccc::cuda::strip_clusterization_algorithm`, etc. I'm quite firm on this...

Reply:
Another way to look at this:

If we **really** wanted to have a single class for describing everything (that I personally don't), then it would be just the `traccc::detector_description` class. It's an SoA. So there's no reason for splitting it into multiple classes. We can just have a bunch of "columns" in it for the different detector types. Then each client would only access the columns that they need.

But again, I'm not actually after such a design. I want to have a clearer separation between different DD types.

Reply:
> But again, I'm not actually after such a design. I want to have a clearer separation between different DD types.

I would also separate them.. just for clarity in the code

Reply:
:confused: There is absolutely no question of intent here. Of course I want to come up with a design that includes wire chamber detectors. I do believe that we have good options for making that happen. I just don't think that your specific proposal would be a good implementation.

But as long as the discussion is about what we would do in the future exactly, we might want to have those discussions in either an appropriate meeting, or in the PR implementing those future changes. :thinking:

Reply:
It's OK as long as you can come up with a design that support the drift chamber and straw tube, and are willing to support them. 

> we could either rename the current class to traccc::sillicon_detector_description, or possibly split it into a traccc::pixel_detector_description and traccc::strip_detector_description classes. And at one point we would introduce a traccc::wire_detector_description class.

Can we do this in this PR?

Reply:
You can leave the wire description totally empty or not even include it. I can work on it later

Reply:
:thinking: I left out many updates from the code on purpose in this PR. Since it's already such a giant that I fear we may never be able to push it in. :frowning: (I'll want to get rid of all of the "old code" most of all. Which this PR doesn't do.)

I mean, I could add a header file called `traccc/geometry/wire_detector_description.hpp` with content like:

```c++
/** TRACCC library, part of the ACTS project (R&D line)
 *
 * (c) 2024 CERN for the benefit of the ACTS project
 *
 * Mozilla Public License Version 2.0
 */

#pragma once

// VecMem include(s).
#include <vecmem/edm/container.hpp>

namespace traccc {

/// Interface for the @c traccc::wire_detector_description class.
///
/// Empty for now.
///
template <typename BASE>
class wire_detector_description_interface : public BASE {};

/// SoA container describing a wire chamber detector
using wire_detector_description = vecmem::edm::container<wire_detector_description_interface>;

}  // namespace traccc
```

But I don't see how that would make anything better in a future PR. :confused: Plus, I'm not actually sure that an SoA container that has no columns whatsoever, would even compile. (In fact I'm pretty sure it wouldn't. Because [vecmem::tuple](https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/utils/tuple.hpp) doesn't allow empty tuples.)

I would really propose to do it in a separate PR...

Reply:
As I said, please do not include it if you don't want, please rename the `detector_desciption` into `pixel_description`or whatever

Reply:
The PR now calls the new class `traccc::silicon_detector_description`.